aiosip/transaction.py

```diff
@@ -37,63 +37,9 @@ def feed_message(self, msg):
-            if self.dialog.password is None:
-                raise ValueError('Password required for authentication')
-
-            if msg.method.upper() == 'REGISTER':
-                self.attempts -= 1
-                if self.attempts < 1:
-                    self.future.set_exception(
-                        RegisterFailed('Too many unauthorized attempts!')
-                    )
-                    return
-                username = msg.to_details['uri']['user']
-            elif msg.method.upper() == 'INVITE':
-                self.attempts -= 1
-                if self.attempts < 1:
-                    self.future.set_exception(
-                        InviteFailed('Too many unauthorized attempts!')
-                    )
-                    return
-                username = msg.from_details['uri']['user']
-
-                hdrs = CIMultiDict()
-                hdrs['From'] = msg.headers['From']
-                hdrs['To'] = msg.headers['To']
-                hdrs['Call-ID'] = msg.headers['Call-ID']
-                hdrs['CSeq'] = msg.headers['CSeq'].replace(self.original_msg.method, 'ACK')
-                hdrs['Via'] = msg.headers['Via']
-                self.dialog.send(method='ACK', headers=hdrs)
-            else:
-                username = msg.from_details['uri']['user']
-
-            del(self.original_msg.headers['CSeq'])
-            self.original_msg.headers['Authorization'] = str(Auth.from_authenticate_header(
-                authenticate=msg.headers['WWW-Authenticate'],
-                method=msg.method,
-                uri=msg.to_details['uri'].short_uri(),
-                username=username,
-                password=self.dialog.password))
-            self.dialog.send(self.original_msg.method,
-                             to_details=msg.to_details,
-                             headers=self.original_msg.headers,
-                             payload=self.original_msg.payload,
-                             future=self.future)
-
-        # for proxy authentication
-        elif msg.status_code == 407:
-            self.original_msg = self.original_msg.pop(msg.cseq)
-            del(self.original_msg.headers['CSeq'])
-            self.original_msg.headers['Proxy-Authorization'] = str(Auth.from_authenticate_header(
-                authenticate=msg.headers['Proxy-Authenticate'],
-                method=msg.method,
-                uri=str(self.to_details),
-                username=self.to_details['uri']['user'],
-                password=self.dialog.password))
-            self.dialog.send_message(msg.method,
-                                     headers=self.original_msg.headers,
-                                     payload=self.original_msg.payload,
-                                     future=self.futrue)
+            self._handle_authenticate(msg)
+        elif msg.status_code == 407:  # Proxy authentication
+            self._handle_proxy_authenticate(msg)
```

Comment:
Oh nice, I started looking at doing this over the weekend... good thing I didn't spend too much time on it 👍

Reply:
So, to expand on this, I started on this to look splitting apart transactions to have proper INVITE/non-INVITE transactions, reliable/unreliable support, and to put in the proper various sip timers and state machines.

I didn't get too far into it on the weekend, and my focus is going to be elsewhere for the most part of this work week. So if you guys need to work on that area, just keep me in the loop to I don't duplicate that effort.

Reply:
Sorry for the false hope, I just separated the methods for Flake8 because it was complaining they were too complex. I don't think it's actually working :/

Reply:
I'm working on the proxy mode, it will probably have a ProxyTransaction or something of this sort

Reply:
Sounds good to me. Yeah, that was the pain I hit too, its a lot of branches to wrap your head around.